pkg/operator/ceph/cluster/osd/status.go

```diff
@@ -278,25 +279,71 @@ func (c *Cluster) findRemovedNodes() (map[string][]*apps.Deployment, error) {
 	// first discover the storage nodes that are still running
 	discoveredNodes, err := c.discoverStorageNodes()
 	if err != nil {
-		return nil, fmt.Errorf("failed to discover storage nodes: %+v", err)
+		return nil, fmt.Errorf("aborting search for removed nodes. failed to discover storage nodes. %+v", err)
+	}
+
+	// c.ValidStorage.Nodes currently in the cluster `c` is only the subset of the user-defined
+	// nodes which are currently valid, and we want a list which can include nodes that are cordoned
+	// for maintenance or have automatic Kubernetes well-known taints added.
+	k8sNodes, err := k8sutil.GetKubernetesNodesMatchingRookNodes(c.DesiredStorage.Nodes, c.context.Clientset)
+	if err != nil {
+		return nil, fmt.Errorf("aborting search for removed nodes. failed to list nodes from Kubernetes. %+v", err)
+	}
+	nodeMap := map[string]v1.Node{}
+	for _, n := range k8sNodes {
+		hostname := n.Labels[apis.LabelHostname]
+		nodeMap[hostname] = n
 	}
 
 	for existingNode, osdDeployments := range discoveredNodes {
-		found := false
-		for _, declaredNode := range c.Storage.Nodes {
-			// discovered storage node still exists in the current storage spec, move on to next discovered node
-			if existingNode == declaredNode.Name {
-				found = true
-				break
-			}
+		var nodeRef *v1.Node
+		if n, ok := nodeMap[existingNode]; !ok {
+			nodeRef = nil
+		} else {
+			nodeRef = &n
 		}
-
-		if !found {
-			// the discovered storage node was not found in the current storage spec, add it to the removed nodes set
-			logger.Infof("adding node %s to the removed nodes list", existingNode)
+		reason := ""
+		if c.nodeRemovedByUser(existingNode, nodeRef) {
+			logger.Infof("adding node %s to the removed nodes list. %s", existingNode, reason)
 			removedNodes[existingNode] = osdDeployments
 		}
 	}
 
 	return removedNodes, nil
 }
+
+// sample of conditions where we cannot determine if the user wants to remove a node as an osd host:
+//  - if the node is not schedulable, it may be cordoned for maintenance
+//  - if the node is not ready, it could be down temporarily
+func (c *Cluster) nodeRemovedByUser(nodeName string, k8sNode *v1.Node) bool {
+	if c.DesiredStorage.UseAllNodes == false {
```

Comment:
Isn't `== false` extraneous? Shouldn't we use `!c.DesiredStorage.UseAllNodes` instead (just like below)?

Reply:
I think it's easier to see `== true` than to see a tiny `!`.